File: app/domain/sequential/tad_lista.py
"""Transcripcion Python de `tad_lista.h`."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Callable


@dataclass
class NodoLista:
    nro: int
    sgte: NodoLista | None = None


Tlista = NodoLista | None

_INSERTAR_RELATIVO_PROVIDER: Callable[[], int] = lambda: 0


def lista_configurar_insertar_antes_despues_provider(provider: Callable[[], int]) -> None:
    """Configura la estrategia para responder `lista_insertar_antes_despues`."""
    global _INSERTAR_RELATIVO_PROVIDER
    _INSERTAR_RELATIVO_PROVIDER = provider


def lista_insertar_inicio(lista: list[Tlista], valor: int) -> None:
    if lista is None:
        return
    q = NodoLista(nro=valor)
    q.sgte = lista[0]
    lista[0] = q


def lista_insertar_final(lista: list[Tlista], valor: int) -> None:
    if lista is None:
        return

    q = NodoLista(nro=valor)
    if lista[0] is None:
        lista[0] = q
        return

    t = lista[0]
    while t.sgte is not None:
        t = t.sgte
    t.sgte = q


def lista_insertar_antes_despues() -> int:
    respuesta = _INSERTAR_RELATIVO_PROVIDER()
    if respuesta not in {-1, 0}:
        return 0
    return respuesta


def lista_insertar_elemento(lista: list[Tlista], valor: int, pos: int) -> None:
    if lista is None or pos <= 0:
        return

    q = NodoLista(nro=valor)
    if pos == 1:
        q.sgte = lista[0]
        lista[0] = q
        return

    desplazamiento = lista_insertar_antes_despues()
    t = lista[0]
    i = 1
    while t is not None:
        if i == pos + desplazamiento:
            q.sgte = t.sgte
            t.sgte = q
            return
        t = t.sgte
        i += 1

    print("   Error...Posicion no encontrada..!")


def lista_buscar_elemento(lista: Tlista, valor: int) -> None:
    i = 1
    encontrado = False
    q = lista

    while q is not None:
        if q.nro == valor:
            print(f"\n Encontrado en la posicion {i}")
            encontrado = True
        q = q.sgte
        i += 1

    if not encontrado:
        print("\n Numero no encontrado..")


def lista_mostrar(lista: Tlista) -> None:
    i = 1
    aux = lista
    while aux is not None:
        print(f" {i}) {aux.nro}")
        aux = aux.sgte
        i += 1


def lista_eliminar_elemento(lista: list[Tlista], valor: int) -> None:
    if lista is None or lista[0] is None:
        print(" Valor no encontrado o lista vacia.")
        return

    p = lista[0]
    ant: NodoLista | None = None

    while p is not None:
        if p.nro == valor:
            if p is lista[0]:
                lista[0] = p.sgte
            else:
                ant.sgte = p.sgte
            return
        ant = p
        p = p.sgte

    print(" Valor no encontrado o lista vacia.")


def lista_eliminar_repetidos(lista: list[Tlista], valor: int) -> None:
    if lista is None or lista[0] is None:
        print("\n\n Valores eliminados..\n")
        return

    q = lista[0]
    ant: NodoLista | None = None

    while q is not None:
        if q.nro == valor:
            if q is lista[0]:
                lista[0] = q.sgte
                q = lista[0]
            else:
                ant.sgte = q.sgte
                q = ant.sgte
        else:
            ant = q
            q = q.sgte

    print("\n\n Valores eliminados..\n")


def lista_buscar_posiciones(lista: Tlista, valor: int) -> list[int]:
    """Helper para la app: devuelve posiciones base 1."""
    posiciones: list[int] = []
    actual = lista
    pos = 1
    while actual is not None:
        if actual.nro == valor:
            posiciones.append(pos)
        actual = actual.sgte
        pos += 1
    return posiciones

File: docs/tads_C/tad_lista.c
#include <stdio.h>
#include <stdlib.h>

/**
 * @brief Nodo de una lista enlazada simple.
 */
struct NodoLista {
    int nro;                     
    struct NodoLista *sgte;      
};

typedef struct NodoLista* Tlista;

static Tlista CrearNodoLista(int valor) {
    Tlista q = (Tlista) malloc(sizeof(struct NodoLista));
    if (q == NULL) {
        printf("Error: no se pudo asignar memoria.\n");
        return NULL;
    }
    q->nro = valor;
    q->sgte = NULL;
    return q;
}


//----------------------------------------------------------------
/**
 * @brief Inserta un nuevo nodo al inicio de la lista.
 * @param lista Puntero a la lista.
 * @param valor Valor entero a insertar.
 */
void lista_insertar_inicio(Tlista *lista, int valor) {
    if (lista == NULL) {
        return;
    }

    Tlista q = CrearNodoLista(valor);
    if (q == NULL) {
        return;
    }

    q->sgte = *lista;
    *lista = q;
}


//---------------------------------------------------------------
/**
 * @brief Inserta un nuevo nodo al final de la lista.
 * @param lista Puntero a la lista.
 * @param valor Valor entero a insertar.
 */
void lista_insertar_final(Tlista *lista, int valor) {
    if (lista == NULL) {
        return;
    }

    Tlista q = CrearNodoLista(valor);
    if (q == NULL) {
        return;
    }

    if (*lista == NULL) {
        *lista = q;
    } else {
        Tlista t = *lista;
        while (t->sgte != NULL) {
            t = t->sgte;
        }
        t->sgte = q;
    }
}

//---------------------------------------------------------------
/**
 * @brief Inserta un nodo en una posición dada.
 * @param lista Puntero a la lista.
 * @param valor Valor a insertar.
 * @param pos Posición base para la inserción.
 */
void lista_insertar_elemento(Tlista *lista, int valor, int pos) {
    if (lista == NULL || pos <= 0) {
        return;
    }

    Tlista q = CrearNodoLista(valor);
    if (q == NULL) {
        return;
    }

    /*
     * Contrato actual:
     * - pos == 1: inserta al inicio.
     * - pos > 1 : inserta despues del nodo en posicion base `pos`.
     */
    if (pos == 1) {
        q->sgte = *lista;
        *lista = q;
        return;
    }

    Tlista t = *lista;
    int i = 1;
    while (t != NULL) {
        if (i == pos) {
            q->sgte = t->sgte;
            t->sgte = q;
            return;
        }
        t = t->sgte;
        i++;
    }

    printf("   Error...Posicion no encontrada..!\n");
    free(q);
}

//---------------------------------------------------------------
/**
 * @brief Busca un elemento en la lista e imprime su posición.
 * @param lista Lista en la que se busca.
 * @param valor Valor a buscar.
 */
void lista_buscar_elemento(Tlista lista, int valor) {
    int i = 1, encontrado = 0;
    Tlista q = lista;

    while (q != NULL) {
        if (q->nro == valor) {
            printf("\n Encontrado en la posicion %d\n", i);
            encontrado = 1;
        }
        q = q->sgte;
        i++;
    }

    if (!encontrado) {
        printf("\n Numero no encontrado..\n");
    }
}


//---------------------------------------------------------------
/**
 * @brief Muestra los elementos de la lista con su posición.
 * @param lista Lista a mostrar.
 */
void lista_mostrar(Tlista lista) {
    int i = 1;
    Tlista aux = lista;
    while (aux != NULL) {
        printf(" %d) %d\n", i, aux->nro);
        aux = aux->sgte;
        i++;
    }
}


//---------------------------------------------------------------
/**
 * @brief Elimina la primera ocurrencia de un valor en la lista.
 * @param lista Puntero a la lista.
 * @param valor Valor a eliminar.
 */
void lista_eliminar_elemento(Tlista *lista, int valor) {
    if (lista == NULL || *lista == NULL) {
        printf(" Valor no encontrado o lista vacia.\n");
        return;
    }

    Tlista p = *lista, ant = NULL;

    while (p != NULL) {
        if (p->nro == valor) {
            if (p == *lista) {
                *lista = p->sgte;
            } else {
                ant->sgte = p->sgte;
            }
            free(p);
            return;
        }
        ant = p;
        p = p->sgte;
    }

    printf(" Valor no encontrado o lista vacia.\n");
}


//---------------------------------------------------------------
/**
 * @brief Elimina todas las ocurrencias de un valor en la lista.
 * @param lista Puntero a la lista.
 * @param valor Valor a eliminar.
 */
void lista_eliminar_repetidos(Tlista *lista, int valor) {
    if (lista == NULL || *lista == NULL) {
        printf("\n\n Valores eliminados..\n");
        return;
    }

    Tlista q = *lista, ant = NULL;

    while (q != NULL) {
        if (q->nro == valor) {
            Tlista temp = q;
            if (q == *lista) {
                *lista = q->sgte;
                q = *lista;
            } else {
                ant->sgte = q->sgte;
                q = ant->sgte;
            }
            free(temp);
        } else {
            ant = q;
            q = q->sgte;
        }
    }
    printf("\n\n Valores eliminados..\n");
}
